Ask interactively for user name and password if Kerberos credentials missing

abbra opened the pull request:
## cli: add acquire_cred callback support to the Kerberos transport

This allows applications to specify their own acquire_cred callback. The callback should be put into the environment under '`acquire_cred`' name as a tuple of a `(callable, pass-in object)`.

The callable should accept `(pass-in object, principal, ccache, service)` and ideally should acquire a Kerberos credential into the `ccache`. Both `principal` and the `ccache` name might be None, signifying that a default name and a default credentials cache would need to be used.

Example usage:
```
import os
from ipalib import api
from ipalib.install.kinit import kinit_password

username = 'admin'
password = '<some-value>'

def acquire_cred(api, principal, ccache, service):
    if ccache is None:
        ccache = os.environ.get('KRB5CCNAME', 'KCM:')
    if principal is None:
        principal = username

    kinit_password(principal, password, ccache)

api.bootstrap(acquire_cred=(acquire_cred, api))
api.finalize()
api.Backend.rpcclient.connect()
print(api.Command.ping())
```

## cli: Ask for user name and password if Kerberos credentials missing

Ask interactively for credentials in case they are missing in the default credentials cache in the environment:
```
	$ ipa ping
	Username: testuser
	Password: <some-password>
	--------------------------------------------
	IPA server version 4.11.1. API version 2.253
	--------------------------------------------
```
Initial implementation only supports password-based credential.

abbra committed:
cli: add acquire_cred callback support to the Kerberos transport

This allows applications to specify their own acquire_cred callback.
The callback should be put into the environment under 'acquire_cred'
name.

The callback would need to adhere to the following signature:

def acquire_cred_cb(
    *,
    api: plugable.API,
    transport: MultiProtocolTransport,
    context: threading.local,
    principal: typing.Optional[str] = None,
    ccache: typing.Optional[str] = None,
    service: typing.Optional[str] = None,
    interactive: bool,
    exc: Exception,
    **kwargs
) -> None:
    """Acquire credential callback

    :param api: ipalib API object
    :param transport: RPC transport instance
    :param context: threading local context with additional data
    :param principal: Kerberos principal name (str, None)
    :param ccache: credential cache with ccache type (str, None)
    :param service: GSSAPI service name (e.g. 'HTTP@host.name' or None)
    :param interactive: true if API is used interactively over TTY
    :param exc: exception to raise in case callback couldn't acquire creds
    """
    pass

and ideally should acquire a Kerberos credential into the ccache. Both
principal and the ccache name might be None, signifying that a default
name and a default credentials cache would need to be used.

Example usage:
-----------------------------------------------
import os
from ipalib import api
from ipalib.install.kinit import kinit_password

username = 'admin'
password = '<some-value>'

def acquire_cred(*, api, principal, ccache, exc, **kw):
    ccache = ccache if ccache else os.environ.get('KRB5CCNAME', 'KCM:')
    principal = principal if principal else username
    try:
        kinit_password(principal, password, ccache,
                       canonicalize=True, enterprise=True)
    except:
        raise exc()

api.bootstrap(acquire_cred=acquire_cred)
api.finalize()
api.Backend.rpcclient.connect()
print(api.Command.ping())
----------------------------------------------

Fixes: https://pagure.io/freeipa/issue/9561

Signed-off-by: Alexander Bokovoy <abokovoy@redhat.com>

abbra committed:
cli: Ask for user name and password if Kerberos credentials missing

Ask interactively for credentials in case they are missing in the
default credentials cache in the environment:

	$ kdestroy -A
        $ ipa ping
	Username[testuser]: admin
	Password: <some-password>
	--------------------------------------------
	IPA server version 4.11.1. API version 2.253
	--------------------------------------------

Only ask for the credentials if we are using a TTY device. This means a
pipe-provided standard input would not trigger ask for credentials:

	$ kdestroy -A
        $ echo -n | ipa ping
        ipa: ERROR: did not receive Kerberos credentials

Initial implementation only supports password-based credential.

Fixes: https://pagure.io/freeipa/issue/9561

Signed-off-by: Alexander Bokovoy <abokovoy@redhat.com>

tiran commented:
I find it confusing how the API object is passed around. How about you include the api in the transport instance and get the callback from `self.api` instead?

```
class MultiProtocolTransport(Transport):
    """Transport that handles both XML-RPC and JSON"""
    def __init__(self, *args, **kwargs):
        Transport.__init__(self)
        self.protocol = kwargs.get('protocol', None)
        self.api = kwargs.get('api', api)
```

and pass the api object along later:

```
                proxy_kw['transport'] = transport_class(
                    protocol=self.protocol, service='HTTP', ccache=ccache,
                    api=self.api)
```

then you can replace the complicated checks with `if self.api.env.acquire_cred is not None`.

The `single_request` method does not pass `service` to `_handle_exception()`.

I think the `acquire_cred_interactively` function won't work for users with 2FA, hardened, or OIDC login. It's not using a FAST tunnel.

How do you feel about a different set of parameters for the callback to make it future-proof?

- require passing of keyword arguments instead of positional arguments
- pass `context` and `transport` in case we need it in the future. We could drop `principal`, `ccache`, and `service` arguments. The information is available on context and transport.
- require `**kwargs` so we can extend the API later without breaking code

```python
import threading
import typing
from ipalib import plugable
from ipalib.rpc import MultiProtocolTransport


def acquire_cred_cb(
    *,
    api: plugable.API,
    transport: MultiProtocolTransport,
    context: threading.local,
    principal: typing.Optional[str] = None,
    ccache: typing.Optional[str] = None,
    service=typing.Optional[str] = None,
    **kwargs
) -> None:
    """Acquire credential callback

    :param api: ipalib API object
    :param transport: RPC transport instance
    :param context: threading local context with additional data
    :param principal: Kerberos principal name (str, None)
    :param ccache: credential cache with ccache type (str, None)
    :param service: service service prefix (e.g. 'HTTP' or None)
    """
    pass
```

tiran commented on ipalib/cli.py:
```diff
@@ -1454,10 +1454,36 @@ def format_description(self, description):
+    from gssapi import Name, NameType
+    from gssapi.raw import (acquire_cred_with_password,
+                            store_cred, store_cred_into)
+    if principal is None:
+        principal = api.Backend.textui.prompt('Username')
+    password = api.Backend.textui.prompt_password('Password', False)
```
You may want to take `api.env.interactive` and `api.env.prompt_all` into account.

abbra replied:
`api.env.in_server` would not be needed to check because this callback will not be called in that case. KerbTransport is only used when we have HTTP transport -- in `in_server` we don't use it.

abbra replied:
I added check for interactivity into `ipalib/rpc.py` and also into the callback. So we wouldn't call into the callback at all in `ipa -n ping` case and will simply fail without creds:
```
ipa: ERROR: did not receive Kerberos credentials
```

abbra commented:
Please note that I explicitly point in the commit message that current implementation will not work for passwordless methods and that is intended -- I am working on an addition to support those but they'll come in a separate PR. This one lays a foundation to enable callbacks that abstract out credentials acquisition at the time when this change happens.

tiran commented:
One place does not pass `service` to `_handle_exception`: https://github.com/freeipa/freeipa/blob/64861a0cf9a8ac18d83a206c11fd3b42be3c578c/ipalib/rpc.py#L736-L737

abbra commented:
Filed a ticket to this specific change: https://pagure.io/freeipa/issue/9561

rcritten commented:
This is a pretty big change from previous behavior. I wonder if in order to retain previous behavior without having to always pass in -n a new config option be added.

tiran commented on ipalib/cli.py:
```diff
@@ -1454,10 +1454,48 @@ def format_description(self, description):
 )
 
 
+def acquire_cred_interactively(*args, **kw):
```
The function should never receive positional arguments and always the keyword arguments api, principal, and ccache, I recommend this syntax:
```suggestion
def acquire_cred_interactively(*, api, principal, ccache, **kwargs):
```

abbra replied:
Thanks, makes sense. It is a nice feature.
I updated the code and also decided to pass in an exception class to raise in case the callback decides it cannot acquire credentials. This helps against loops within `self.get_auth_info()`: `self._handle_exception()` -> `self._try_acquire_cred()` -> `raise errors.NonFatalError()` -> `self.get_auth_info()`.

tiran commented on ipalib/rpc.py:
```diff
@@ -586,12 +587,31 @@ def __init__(self, *args, **kwargs):
         self.ccache = kwargs.pop("ccache", None)
 
     def _handle_exception(self, e, service=None):
+        def try_acquire_cred():
```
IMHO the code would be more readable if you move the function from local scope to a private method `_try_acquire_cred`.

abbra replied:
Thanks, addressed.

tiran reviewed (commented):
I see another potential problem with `api.env.interactive`. It's always set to `True` unless `ipa` CLI is called with `ipa --no-prompt`. This may cause issues with scenarios like systemd services or Ansible playbooks that run `ipa` non-interactively. I haven't tested it, but I think that the new code is going to block on stdin read indefinitely.

## Suggestions:

For password prompt, only use the interactive flag if stdin is a tty:
```python
interactive = (self.api.env.interactive and sys.stdin.isatty()) or self.api.env.prompt_all
```

then pass `interactive` to the `acquire_cred` callback. This would let each callback implementation detect if interaction with a user is possible. The reference implementation of `acquire_cred_interactively` does nothing, but a callback with a password from a config file would work.

abbra committed:
rpc: reuse GSSAPI service name through an exception handler

When we encounter GSSAPI exception, it means we are either in a process
of negotiating authentication with a remote service or we just started
acquiring a ticket. In either case, we should know GSSAPI service name:
either it was passed to us by get_auth_info() directly or can infer it
from the opened connection.

Fixes: https://pagure.io/freeipa/issue/9561

Signed-off-by: Alexander Bokovoy <abokovoy@redhat.com>

abbra committed:
ipapython/session_storage.py: pylint suggestions

Signed-off-by: Alexander Bokovoy <abokovoy@redhat.com>

abbra committed:
makeapi.in: update to follow Python 3.11+ changes and reformat

./makeapi:443: SyntaxWarning: invalid escape sequence '\('
  m = re.match('^[a-zA-Z0-9]+\(\'([a-z][_a-z0-9?\*\+]*)\'.*', line)

Changing just one line will cause surrounding lines to be affected by
the Pylint checks. Changing those lines will expose more code to the
checks. In the end, I had to reformat the whole script.

Signed-off-by: Alexander Bokovoy <abokovoy@redhat.com>

abbra committed:
ipapython/kerberos: move low level Kerberos code to ipapython.kerberos

In order to add bindings to more MIT Kerberos functions, move existing
ones to a separate file. Session code uses few of the Kerberos functions
to operate on the credential caches.

Signed-off-by: Alexander Bokovoy <abokovoy@redhat.com>

abbra committed:
ipapython/kerberos: add initial bindings to allow kinit implementation

Initializing credentials with passwordless pre-authentication methods
requires two major elements:

 - use of FAST channel for exposing passwordless methods
 - use of pre-authentication method prompts

Communicating with kinit command line utility to provide prompts is
awkward. Instead, expose MIT Kerberos API for prompting.

This commit adds initial kinit_with_cb() implementation that allows to
invoke an externally provided callback as a prompter. A default callback
implementation will be provided later.

Signed-off-by: Alexander Bokovoy <abokovoy@redhat.com>

abbra committed:
ipalib/kinit: initial implementation of interactive prompter

Add initial implementation of the interactive prompter to be used with
ipapython.kerberos.kinit_with_cb(). The prompter can work with all
preauthentication methods FreeIPA supports that require prompts:

 - password-based ones (timestamp and SPAKE)
 - OTP/RADIUS
 - passkey
 - external IdP

Signed-off-by: Alexander Bokovoy <abokovoy@redhat.com>

abbra commented:
I am turning this back to draft mode. I have made some progress with interactive prompter to get rid of calling out to `kinit`, but haven't yet integrated it into the helpers for automatic use.

abbra commented:
This PR is in works.
